golem-component-service-base/src/repo/component.rs

```diff
@@ -0,0 +1,510 @@
+            protected_component: component.protected_component_id.slug(),
+            protector_version: None,
+            metadata: metadata.into(),
+        })
+    }
+}
+
+#[async_trait]
+pub trait ComponentRepo {
+    async fn upsert(&self, component: &ComponentRecord) -> Result<(), RepoError>;
+
+    async fn get(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_all(&self, namespace: &str) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_latest_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Option<ComponentRecord>, RepoError>;
+
+    async fn get_by_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+        version: u64,
+    ) -> Result<Option<ComponentRecord>, RepoError>;
+
+    async fn get_by_name(
+        &self,
+        namespace: &str,
+        name: &str,
+    ) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_ids_by_name(&self, namespace: &str, name: &str) -> Result<Vec<Uuid>, RepoError>;
+
+    async fn get_namespaces(&self, component_id: &Uuid) -> Result<Vec<String>, RepoError>;
+
+    async fn delete(&self, namespace: &str, component_id: &Uuid) -> Result<(), RepoError>;
+}
+
+pub struct DbComponentRepo<DB: Database> {
+    db_pool: Arc<Pool<DB>>,
+}
+
+impl<DB: Database> DbComponentRepo<DB> {
+    pub fn new(db_pool: Arc<Pool<DB>>) -> Self {
+        Self { db_pool }
+    }
+}
+
+#[async_trait]
+impl ComponentRepo for DbComponentRepo<sqlx::Sqlite> {
+    async fn upsert(&self, component: &ComponentRecord) -> Result<(), RepoError> {
+        let mut transaction = self.db_pool.begin().await?;
+
+        let result = sqlx::query(
+            "SELECT count(component_id) as count FROM components WHERE namespace != $1 AND component_id = $2",
+        )
+        .bind(component.namespace.clone())
+        .bind(component.component_id)
+        .fetch_one(&mut *transaction)
+        .await?;
+
+        let count: i64 = result.get("count");
+
+        if count > 0 {
+            return Err(RepoError::Internal("Component id not unique".to_string()));
+        }
+
+        sqlx::query(
+            r#"
+              INSERT INTO components
+                (namespace, component_id, name)
+              VALUES
+                ($1, $2, $3)
+              ON CONFLICT (namespace, name) DO NOTHING
+               "#,
+        )
+        .bind(component.namespace.clone())
+        .bind(component.component_id)
+        .bind(component.name.clone())
+        .execute(&mut *transaction)
+        .await?;
+
+        sqlx::query(
+            r#"
+              INSERT INTO component_versions
+                (component_id, version, size, user_component, protected_component, protector_version, metadata)
+              VALUES
+                ($1, $2, $3, $4, $5, $6, $7)
+              ON CONFLICT (component_id, version) DO UPDATE
```

Comment:
Why do we have a version update at all? Should we keep it immutable?
What is the meaning of updating the size? We can't just replace component version code.

Reply:
it was upsert before, but I think you have good point, it should be just insert ..., i changed it to insert